### PR TITLE
Use new UtxoEnv type in Alonzo

### PR DESCRIPTION
**Pull request by JaredCorduan**
In alonzo, the `UTxOS` rule needs to be able to resolve pointer address in order to pass the stake credentials to the plutus interpreter. Therefor the `UtxoEnv` type needs the pointer map.

This was nearly a small change, except that the `UTXOW` rule in all eras uses the `shelleyStyleWitness` function, which was previous fixed to the `UtxoEnv` type. The `shelleyStyleWitness` function is now parametric in this type, but this came at the cost of `applyBlock` and `applyTxs` now needing this type to be specified.

Is this a good approach? There are other options, but they don't seem great to me. We could just add the pointer map to the type for all eras, and not use it pre-alonzo. We could also not attempt to use `shelleyStyleWitness` in alonzo, but this seems sad too. We could also introduce another type family.

### CODE REVIEW

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs [hunk 1]
```diff
@@ -33,18 +34,31 @@ import GHC.Records (HasField (..))
...
+import Shelley.Spec.Ledger.TxBody (DCert, PoolParams, Ptr, TxIn (..), Wdrl)
 import Shelley.Spec.Ledger.UTxO (balance, totalDeposits)
 
 --------------------------------------------------------------------------------
 -- The UTXOS transition system
 --------------------------------------------------------------------------------
 
+-- | The UTXO Environment in the Alonzo Era
+data UtxoEnv era = UtxoEnv
+  { slotUE :: SlotNo,
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:53]: This postfix type descriptor is unusual, and I don't think we use it anywhere else. So would tend to prefer e.g. `ueSlot`

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API.hs [hunk 2]
```diff
@@ -46,6 +47,6 @@ class
     ChainData (State (Core.EraRule "PPUP" era)),
     SerialisableData (State (Core.EraRule "PPUP" era))
   ) =>
-  ShelleyBasedEra era
+  ShelleyBasedEra era utxoenv
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API.hs:50]: I don't think this is going to work. Consensus uses this to cover all shelley based ledgers, but this parameter will have to vary.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs [hunk 3]
```diff
@@ -252,7 +251,7 @@ utxoTransition ::
     ValidateScript era,
     -- instructions for calling UTXOS from AlonzoUTXO
     Embed (Core.EraRule "UTXOS" era) (AlonzoUTXO era),
-    Environment (Core.EraRule "UTXOS" era) ~ Shelley.UtxoEnv era,
+    Environment (Core.EraRule "UTXOS" era) ~ UtxoEnv era,
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:254]: One thing we can do here is simply bind the two type families together:

`Environment (Core.EraRule "UTXOS" era) ~ Environment (Core.EraRule "UTXO" era)`

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs [hunk 4]
```diff
@@ -63,9 +68,11 @@ class
     BaseM (Core.EraRule "BBODY" era) ~ ShelleyBase,
     Environment (Core.EraRule "BBODY" era) ~ STS.BbodyEnv era,
     State (Core.EraRule "BBODY" era) ~ STS.BbodyState era,
-    Signal (Core.EraRule "BBODY" era) ~ Block era
+    Signal (Core.EraRule "BBODY" era) ~ Block era,
+    HasField "pparamsUE" (utxoenv era) (Core.PParams era),
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:72]: Why not simply `HasField "pparamsUE" (Environment (Core.EraRule "UTXO" era)) (Core.PParams era)`? That avoids the need for this extra param.

**Review by nc6 — COMMENTED** [review]
I don't think this will easily work, but I've tried to suggest a way that might. Can discuss later!

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API.hs (hunk 2 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API.hs:50]: ok, that's what I was worried about, thanks!

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs (hunk 3 above)
Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:254]: oh, interesting! thanks!

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs (hunk 4 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:72]: great!

**Comment by JaredCorduan** [comment]
I made all of @nc6 's wonderful suggestions, it cleaned things up very nicely! But we no longer want these changes, we are _not_ going to resolve pointer addresses for plutus, this would break determinism. Plutus scripts will just have to use base addresses if they want this kind of information.

**Pull request closed without merging**